spark/src/test/scala/com/gu/tableversions/spark/SparkHiveMetastoreSpec.scala

```diff
@@ -0,0 +1,138 @@
+package com.gu.tableversions.spark
+
+import cats.effect.IO
+import cats.syntax.functor._
+import com.gu.tableversions.core.Partition.{ColumnValue, PartitionColumn}
+import com.gu.tableversions.core._
+import com.gu.tableversions.metastore.MetastoreSpec
+import org.scalatest.prop.PropertyChecks
+import org.scalatest.{FlatSpec, Matchers}
+import SparkHiveMetastore._
+
+class SparkHiveMetastoreSpec extends FlatSpec with Matchers with SparkHiveSuite with MetastoreSpec with PropertyChecks {
+
+  val snapshotTable =
+    TableDefinition(TableName(schema, "users"), resolveTablePath("users"), PartitionSchema.snapshot)
+
+  val partitionedTable = TableDefinition(TableName(schema, "clicks"),
+                                         resolveTablePath("clicks"),
+                                         PartitionSchema(List(PartitionColumn("date"))))
+  //
+  // Common specs
+  //
+
+  "A metastore implementation" should behave like metastoreWithSnapshotSupport(IO { new SparkHiveMetastore },
+                                                                               initSnapshotTable(snapshotTable),
+                                                                               snapshotTable)
+
+  it should behave like metastoreWithPartitionsSupport(IO { new SparkHiveMetastore },
+                                                       initPartitionedTable(partitionedTable),
+                                                       partitionedTable)
+
+  //
+  // Tests specific to the Spark/Hive implementation
+  //
+
+  "Parsing a valid partition string" should "produce the expected values" in {
+    val testData = Table(
+      ("partitionStr", "expected"),
+      ("date=2019-01-31", Partition(List(ColumnValue(PartitionColumn("date"), "2019-01-31")))),
+      ("event_date=2019-01-30/processed_date=2019-01-31",
+       Partition(
+         List(ColumnValue(PartitionColumn("event_date"), "2019-01-30"),
+              ColumnValue(PartitionColumn("processed_date"), "2019-01-31")))),
+      ("year=2019/month=01/day=31",
+       Partition(
+         List(ColumnValue(PartitionColumn("year"), "2019"),
+              ColumnValue(PartitionColumn("month"), "01"),
+              ColumnValue(PartitionColumn("day"), "31"))))
+    )
+
+    forAll(testData) { (partitionStr: String, expected: Partition) =>
+      SparkHiveMetastore.parsePartition(partitionStr) shouldBe expected
+    }
+  }
+
+  "Parsing an invalid partition string" should "throw an exception" in {
+    // format: off
+    val invalidPartitionStrings =
+      Table(
+        "partitionString",
+        "invalid partition string",
+        "invalid partition string=42",
+        //"/",
+        "")
+    // format: on
+
+    forAll(invalidPartitionStrings) { partitionStr =>
+      an[Exception] should be thrownBy SparkHiveMetastore.parsePartition(partitionStr)
+    }
+  }
+
+  "Rendering a partition as a string" should "produce a valid Hive partition expression" in {
```

Comment:
Even if we don't use ScalaCheck, there's a round-trip property here that might be worth testing explicitly.

Maybe a single-column `Table` containing Hive partition expressions that when parsed and then serialised is unchanged.

Reply:
Roundtripping would be nice, but we don't have any code for parsing Hive partition expressions, only partition paths.

The former is expressions like `(event_date='2019-01-30', processed_date='2019-01-31')`, which is the syntax used in queries for adding partitions. I.e. we only render such strings for use in queries, we never parse them.

The latter is strings like `event_date=2019-01-30/processed_date=2019-01-31`, which is what we get back from "show partitions" queries. These ones we only parse when they come back, we don't produce them.